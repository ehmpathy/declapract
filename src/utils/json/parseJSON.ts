import JSON5 from 'json5';

export class JSONParsingError extends Error {
  constructor({
    jsonError,
    json5Error,
    contents,
  }: {
    jsonError: Error;
    json5Error: Error;
    contents: string;
  }) {
    super(
      `
Could not parse contents as neither JSON nor JSON5.

JSON Error:
${jsonError.message}

JSON5 Error:
${json5Error.message}

Contents Head:
${contents.slice(0, 50)}${contents.length > 50 ? '...' : ''}
    `.trim(),
    );
  }
}

export enum JSONVariant {
  JSON = 'JSON',
  // eslint-disable-next-line @typescript-eslint/no-shadow
  JSON5 = 'JSON5',
}

/**
 * parse json with support for json5, when needed
 */
export const parseJSON = <T>(
  contents: string,
): { output: T; variant: JSONVariant } => {
  try {
    return { variant: JSONVariant.JSON, output: JSON.parse(contents) };
  } catch (jsonError) {
    try {
      return { variant: JSONVariant.JSON5, output: JSON5.parse(contents) };
    } catch (json5Error) {
      throw new JSONParsingError({ json5Error, jsonError, contents });
    }
  }
};
