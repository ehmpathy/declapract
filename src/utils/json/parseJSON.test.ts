import { JSONParsingError, parseJSON } from './parseJSON';

describe('parseJSON', () => {
  it('should be able to parse normal json', () => {
    const original = { hello: 'world' };
    const output = parseJSON(JSON.stringify(original));
    expect(output).toEqual(original);
  });
  it('should be able to parse json5', () => {
    const input = `{"hello": "world", }`; // note the trailing comma
    const output = parseJSON(input);
    expect(output).toEqual({ hello: 'world' });
  });
  it('should throw an error if neither is able to parse', () => {
    const input = `{"hello": z "world", }`; // note the random z
    try {
      parseJSON(input);
      throw new Error('should not reach here');
    } catch (error) {
      expect(error).toBeInstanceOf(JSONParsingError);
      expect(error.message).toContain(
        `Could not parse contents as neither JSON nor JSON5.`,
      );
    }
  });
});
