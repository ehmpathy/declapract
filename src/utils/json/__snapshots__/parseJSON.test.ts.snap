// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseJSON should throw an error if neither is able to parse 1`] = `
"Could not parse contents as neither JSON nor JSON5.

JSON Error:
Unexpected token z in JSON at position 10

JSON5 Error:
JSON5: invalid character 'z' at 1:11

Contents Head:
{"hello": z "world", }"
`;
